internal/path/segment_trie/segment_trie.go

```diff
@@ -0,0 +1,167 @@
+//   - Nodes that are pointed to by `{**}` don't store their children,
+//     instead they store the path suffix that exists after `{**}`.
+//     Conflicts are checked by comparing the suffixes of paths with the same segments that precede `{**}`.
+//     This can be done, since the `{**}` operator must be the last in the path.
+//   - `{*}` nodes are stored just like any other exact segment node, but are always included in conflict search.
+package segment_trie
+
+import (
+	"errors"
+	"github.com/kyma-project/api-gateway/internal/path/token"
+	"strings"
+)
+
+type SegmentTrie struct {
+	Root *Node
+}
+
+type Node struct {
+	EndNode  bool             `json:"-"`
+	Children map[string]*Node `json:"children"`
+	Suffixes []string         `json:"suffixes"`
+}
+
+func New() *SegmentTrie {
+	return &SegmentTrie{
+		Root: &Node{
+			EndNode:  false,
+			Children: map[string]*Node{},
+		},
+	}
+}
+
+func (t *SegmentTrie) InsertAndCheckCollisions(tokens []token.Token) error {
+	if len(tokens) == 0 {
+		return nil
+	}
+
+	node := t.Root
+	if len(t.Root.Children) != 0 {
+		pathExist := findExistingPath(node, tokens, 0)
+		if pathExist {
+			return errors.New("path collision detected")
+		}
+	}
+
+	for i, tok := range tokens {
+		if tok.Type == token.BracedDoubleAsterix {
+			if _, ok := node.Children["{**}"]; !ok {
+				node.Children["{**}"] = &Node{
+					EndNode:  i == len(tokens)-1,
+					Children: make(map[string]*Node),
+					Suffixes: []string{token.List(tokens[i+1:]).String()},
+				}
+			} else {
+				suffixString := token.List(tokens[i:]).String()
+				node.Children["{**}"].Suffixes = append(node.Children["{**}"].Suffixes, suffixString)
+			}
+		} else {
+			if _, ok := node.Children[tok.Literal]; !ok {
+				node.Children[tok.Literal] = &Node{
+					EndNode:  i == len(tokens)-1,
+					Children: make(map[string]*Node),
+				}
+			}
+		}
+
+		node = node.Children[tok.Literal]
+	}
+	return nil
+}
+
+func suffixExist(node *Node, suffix []token.Token, cur int) bool {
+	if cur >= len(suffix) {
+		return true
+	}
+
+	if cnode, ok := node.Children["{**}"]; ok {
+		tokensString := token.List(suffix).String()
+		for _, suffix := range cnode.Suffixes {
+			if strings.HasSuffix(tokensString, suffix) || strings.HasSuffix(suffix, tokensString) {
+				return true
+			}
+		}
+	}
+
+	if _, ok := node.Children[suffix[cur].Literal]; ok {
+		if suffixExist(node.Children[suffix[cur].Literal], suffix, cur+1) {
+			return true
+		}
+	}
+
+	for k, v := range node.Children {
+		if k != "{**}" && suffixExist(v, suffix, cur) {
+			return true
+		}
+	}
+	return false
+}
+
+func findExistingPath(node *Node, tokens []token.Token, cur int) bool {
+	if len(tokens) == 0 {
+		return node.EndNode
+	}
+
+	if cur >= len(tokens) {
+		return node.EndNode
+	}
+
+	if len(node.Suffixes) > 0 {
+		return hasAnySuffix(tokens, node.Suffixes)
+	}
+
+	switch tokens[cur].Type {
+	case token.Ident:
+		if _, ok := node.Children[tokens[cur].Literal]; ok {
```

Comment:
You reference the *value* of element you checked existence. Why don't you just use this value? This suggestion also applies to the IFs below.
```suggesion
		if val, ok := node.Children[tokens[cur].Literal]; ok {
					if findExistingPath(val, tokens, cur+1) {
				return true
			}}
```

I also don't like the nested value reference. I think for better readability you *should* assign current token you work with as `token := tokens[cur]` and then reference created variable in switch statement.